File: src-tauri/src/commands/courses.rs
use crate::db::courses::{self, CourseGrades, Semester};
use crate::db::entities::Entity;
use crate::db::DbState;
use crate::error::AppResult;
use tauri::State;

#[tauri::command]
pub fn create_course(state: State<DbState>, space_id: String, title: String) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    courses::create_course(&conn, space_id, title)
}

#[tauri::command]
pub fn list_courses(state: State<DbState>, space_id: String) -> AppResult<Vec<Entity>> {
    let conn = state.0.lock().unwrap();
    courses::list_courses(&conn, &space_id)
}

#[tauri::command]
pub fn create_semester(
    state: State<DbState>,
    space_id: String,
    title: String,
    start_date: Option<String>,
    end_date: Option<String>,
    term_type: Option<String>,
    year: Option<i64>,
) -> AppResult<Semester> {
    let conn = state.0.lock().unwrap();
    courses::create_semester(
        &conn, space_id, title, start_date, end_date, term_type, year,
    )
}

#[tauri::command]
pub fn list_semesters(state: State<DbState>, space_id: String) -> AppResult<Vec<Semester>> {
    let conn = state.0.lock().unwrap();
    courses::list_semesters(&conn, &space_id)
}

#[tauri::command]
pub fn update_semester(
    state: State<DbState>,
    entity_id: String,
    start_date: Option<String>,
    end_date: Option<String>,
    term_type: Option<String>,
    year: Option<i64>,
) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    courses::update_semester(&conn, &entity_id, start_date, end_date, term_type, year)
}

#[tauri::command]
pub fn set_current_semester(
    state: State<DbState>,
    space_id: String,
    entity_id: String,
) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    courses::set_current_semester(&conn, &space_id, &entity_id)
}

#[tauri::command]
pub fn reorder_semesters(state: State<DbState>, ordered_ids: Vec<String>) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    courses::reorder_semesters(&conn, ordered_ids)
}

#[tauri::command]
pub fn link_course_to_semester(
    state: State<DbState>,
    course_id: String,
    semester_id: String,
) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    courses::link_course_to_semester(&conn, course_id, semester_id)
}

#[tauri::command]
pub fn set_course_semester(
    state: State<DbState>,
    course_id: String,
    semester_id: String,
) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    courses::set_course_semester(&conn, &course_id, semester_id)
}

#[tauri::command]
pub fn get_course_notes(state: State<DbState>, course_id: String) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    courses::get_or_create_course_notes(&conn, &course_id)
}

#[tauri::command]
pub fn get_semester_notes(state: State<DbState>, semester_id: String) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    courses::get_or_create_semester_notes(&conn, &semester_id)
}

#[tauri::command]
pub fn get_course_grades(state: State<DbState>, course_id: String) -> AppResult<CourseGrades> {
    let conn = state.0.lock().unwrap();
    courses::get_course_grades(&conn, &course_id)
}

File: packages/frontend/src/lib/api/courses.ts
import { invoke } from "@tauri-apps/api/core";
import type { CourseGrades, Entity, Semester } from "./types";

export function createCourse(spaceId: string, title: string): Promise<Entity> {
  return invoke("create_course", { spaceId, title });
}

export function listCourses(spaceId: string): Promise<Entity[]> {
  return invoke("list_courses", { spaceId });
}

export function createSemester(
  spaceId: string,
  title: string,
  opts: {
    startDate?: string | null;
    endDate?: string | null;
    termType?: string | null;
    year?: number | null;
  } = {},
): Promise<Semester> {
  return invoke("create_semester", {
    spaceId,
    title,
    startDate: opts.startDate ?? null,
    endDate: opts.endDate ?? null,
    termType: opts.termType ?? null,
    year: opts.year ?? null,
  });
}

export function listSemesters(spaceId: string): Promise<Semester[]> {
  return invoke("list_semesters", { spaceId });
}

export function updateSemester(
  entityId: string,
  patch: { startDate?: string; endDate?: string; termType?: string; year?: number },
): Promise<void> {
  return invoke("update_semester", {
    entityId,
    startDate: patch.startDate ?? null,
    endDate: patch.endDate ?? null,
    termType: patch.termType ?? null,
    year: patch.year ?? null,
  });
}

/// Manual "this is the current semester" override — clears the flag on
/// every other Semester in the Space (data layer enforces at most one).
export function setCurrentSemester(spaceId: string, entityId: string): Promise<void> {
  return invoke("set_current_semester", { spaceId, entityId });
}

/// Persists a drag-reorder — `orderedIds` is the full new display order.
export function reorderSemesters(orderedIds: string[]): Promise<void> {
  return invoke("reorder_semesters", { orderedIds });
}

/// A Course has at most one Semester — this errors if the Course is already
/// linked to a different one. Use `setCourseSemester` to reassign instead.
export function linkCourseToSemester(courseId: string, semesterId: string): Promise<void> {
  return invoke("link_course_to_semester", { courseId, semesterId });
}

/// Assigns/reassigns a Course's Semester, dropping any existing link first —
/// always safe to call whether or not the Course already has one.
export function setCourseSemester(courseId: string, semesterId: string): Promise<void> {
  return invoke("set_course_semester", { courseId, semesterId });
}

/// Finds this Course's auto-created Course Notes page, creating it on first
/// request if one doesn't exist yet (e.g. a Course created before this feature).
export function getCourseNotes(courseId: string): Promise<Entity> {
  return invoke("get_course_notes", { courseId });
}

/// Finds this Semester's auto-created Notes page (a real `note` entity,
/// unlike Course Notes — see `get_or_create_semester_notes` in courses.rs),
/// creating it on first request if one doesn't exist yet.
export function getSemesterNotes(semesterId: string): Promise<Entity> {
  return invoke("get_semester_notes", { semesterId });
}

export function getCourseGrades(courseId: string): Promise<CourseGrades> {
  return invoke("get_course_grades", { courseId });
}
